src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,288 @@
+use crate::{
+    chain::chain_information::{
+        babe_fetch_epoch, BabeEpochInformation, ChainInformation, ChainInformationConsensus,
+        ChainInformationFinality,
+    },
+    executor::{
+        host::{HostVmPrototype, NewErr},
+        vm::ExecHint,
+    },
+    finality::{grandpa::warp_sync, justification::verify},
+    header::Header,
+    libp2p::PeerId,
+    network::protocol::GrandpaWarpSyncResponseFragment,
+};
+
+/// Problem encountered during a call to [`grandpa_warp_sync`].
+#[derive(Debug, derive_more::Display)]
+pub enum Error {
+    #[display(fmt = "All requests failed.")]
+    AllRequestsFailed,
+    #[display(fmt = "{}", _0)]
+    Verifier(verify::Error),
+    #[display(fmt = "{}", _0)]
+    BabeFetchEpoch(babe_fetch_epoch::Error),
+    #[display(fmt = "{}", _0)]
+    NewRuntime(NewErr),
+}
+
+/// The configuration for [`grandpa_warp_sync`].
+pub struct Config<'a> {
+    /// A list of connected peers.
+    pub connected_peers: Vec<PeerId>,
```

Comment:
```suggestion
```

We need to be able to add and remove peers while the Grandpa warp syncing is running.

I think that `connected_peers` shouldn't be passed in the config.
Instead:
- The `GrandpaWarpSync` object should start in some sort of `WaitingForPeers` variant.
- Add `WarpSyncRequest::add_peer` and `WarpSyncRequest::remove_other_peer` (`remove_other_peer` should panic if you remove the `current_peer`).
- Instead of returning `AllRequestsFailed` if all requests have failed, switch back to `WaitingForPeers`.